# Bioinformatics Workshop

Welcome to the Fall 2023 and Spring 2024 Bioinformatics (BFX) Workshop! 


## Objectives

This workshop series is designed for people who want to:

- learn the fundamentals of computational genomics
- use this information to improve and expedite their research
- expand their knowledge by hearing from experts in other subject areas

By the end of this course, participants will be able to:

- Demonstrate basic competency in running key bioinformatics software
- Analyze sequencing data, going from raw data to biologically interpretable results
- Draw upon a diverse set of genomic technologies to answer specific research questions
- Apply critical thinking and analytic competencies to genomics research

## Registration

You are welcome to join at any time by [registering on this form](https://redcap.link/BFX2023). After registering, you'll be added to an email distribution list and slack channel for the course. See the [Communications](README.md#Communications) section below for details. There is no obligation to complete the course if you register, but it helps us with planning and securing appropriate resources (e.g. cloud computing accounts)

## Schedule

Weekly Workshops on Mondays @ 10am in [Mid-Campus Center](https://maps.google.com/?q=4590%20Children) (MCC) 11th floor Main Conference Room (Room #11504)

The BFX Workshop schedule, __subject to change__, is available below. Previous years' course materials can still be accessed in the [archive](archive) folder. We *strongly* encourage in-person participation, but a Zoom recording of each Workshop session will also be made available shortly after each session. See [Participation](README.md#Participation) for more details about external or remote participation.

Time will be set aside before and after lectures for "office hours", where you can get help with assignments, or ask for advice about your own projects.

### Fall 2023
|Date|Topic|Presenter|
|----|--------|------------|
|09/11/23	| [Workshop Intro and Prerequisite Review](lectures/week_01) | Chris Miller |
|09/18/23	| [Command line skills, Basic plotting with R](lectures/week_02) |Chris Miller |
|09/25/23	| [Sequence data generation/manipulation](lectures/week_03) | Chris Miller |
|10/2/23	| [DNA Alignment Fundamentals](lectures/week_04) | Jason Walker |
|10/09/23	| NO SEMINAR – FALL BREAK	| |
|10/16/23	| [Sequence data visualization/IGV](lectures/week_05) | Chris Miller |
|10/23/23	| [Germline and Somatic Variant Calling](lectures/week_06) | Chris Miller |
|10/30/23	| [Parsing, Filtering, and Annotating VCFs](lectures/week_07) | Susanna Kiwala, John Garza |
|11/06/23	| [Data visualization with R and ggplot2](lectures/week_08) | Chris Miller |
|11/13/23	| NO SEMINAR  | | 
|11/20/23 | [Long Read Sequencing](lectures/week_09)	| Chris Miller |
|11/27/23	| [RNA sequencing I – Data and Alignment](lectures/week_10) | Kartik Singhal |
|12/04/23	| [RNA sequencing II – Transcript Quantification](lectures/week_11) | My Hoang |
|12/11/23	| [RNA sequencing III – Differential Expression](lectures/week_12) | Mariam Khanfar |

### Spring 2024
|Date|Topic|Presenter|
|----|--------|------------|
| 01/22/24 | [Course review, prereqs, bedtools](lectures/week_13) | Chris Miller |
| 01/29/24 | [Epigenomics, ChIP/ATAC/WGBS](lectures/week_14) | Chris Miller |
| 02/05/24 | [Data visualization](lectures/week_15) | Chris Miller |
| 02/12/24 | [Single-cell RNAseq part 1](lectures/week_16) | Jennifer Foltz |
| 02/19/24 | [Single-cell RNAseq part 2](lectures/week_17) | Jennifer Foltz |
| 02/26/24 | [When Good Experiments Go Bad](lectures/week_18) | Chris Miller |
| 03/4/24 | [Genome Assembly, Pangenome](lectures/week_19) | Juan Macias |
| 03/11/24 | NO SEMINAR - SPRING BREAK | NO SEMINAR - SPRING BREAK |
| 03/18/24 | [ctDNA/error-corrected sequencing](lectures/week_20) | Chris Miller |
| 03/25/24 | NO SEMINAR |  |
| 04/1/24 | [Microbial Genomics](lectures/week_21) | Brigida Rusconi |
| 04/8/24 | NO SEMINAR |   |
| 04/15/24 | [Cloud Computing part 1](lectures/week_22) | Malachi Griffith |
| 04/22/24 | [Genomic Workflows](lectures/week_23) | Jason Walker |
| 04/29/24 | Genomic Medicine | Chris Miller |



## Participation

There are no requirements or qualifications to participate in the BFX Workshop. In order to be successful, you will have to work through the [Prerequisites](lectures/week_01/bfx_workshop_01_overview.ipynb) tutorial to understand the computer systems, tools, and languages that will be used throughout the course. Most of the languages and tools used are common to all fields and applications of Bioinformatics. No academic credits are earned for attending the course and there no attendance requirements. There will be "assignments" given during most lectures and we strongly encourage you to complete them in order to get the most out of the course - there is no substitute for hands-on learning!

For external or remote asyncronous participation, we suggest watching each weekly Zoom recording (available shortly after each in-person session) and strongly encourage interaction via [Slack](README.md#Slack). The #bfx_workshop Slack channel is the primary form of interaction between workshop attendees, external participants, organizers, and presenters. 

## Organizers

This workshop is a collaborative effort between the Institute for Clinical and Translational Sciences (ICTS), McDonnell Genome Institute (MGI), the Department of Medicine, the Division of Hematology & Oncology, and the Department of Pediatrics with faculty, staff, and student representation and participation from many other Institutes, Departments, and Divisions at Washington University School of Medicine in St. Louis.

The organizers would like to specifically thank the [ICTS Precision Health](https://icts-precisionhealth.wustl.edu/) function for supporting the Bioinformatics Workshop. NIH CTSA Grant Number UL1TR002345

The organizers and presenters of this workshop have multi-disciplinary backgrounds in Bioinformatics, Medicine, Computational Biology, Genetics, Computer Science, Oncology, Virology, Information Systems, Microbiology, Immunology, and many other combinations of both Data and Biological Sciences.

This team of organizers and presenters are committed to sharing best practice methods, technology, and knowledge from decades of experience working in these fields of study. New and novel applications of genome sequencing data are constantly evolving and even more so the analytical processes and tools used to translate data to information and knowledge are ever growing and changing. The team leading this workshop will focus on standard best practices while having an eye toward the future direction of each Bioinformatics topic.

## Communications

### Slack

Announcements, discussion, and homework help all take place primarily via Slack. For members of the ICTS Precision Health Slack organization [ictsprecisionhealth.slack.com](http://ictsprecisionhealth.slack.com), please use the [#bfx-workshop](https://ictsprecisionhealth.slack.com/archives/C040Q704WS2) channel. If you do not have access to the ICTS Precision Health Slack, please email [j.mckenzie at wustl.edu]

### Email

An Office 365 Group will be created for general announcements, schedule changes, etc from those that [register](README.md#Registration).  

### Archived Lectures

If you'd like to look back at past sessions, course info and lectures from previous academic years are available in the archives:

- [2020-2021](archive/v2020-2021)
- [2021-2022](archive/v2021-2022) 
- [2022-2023](archive/v2022-2023)
