---
title: "BFX_2025_Day_1"
author: "Jennifer Foltz"
date: "2025-02-06"
output: html_document
params:
  res: 0.8 # where this is the resolution for your clustering
  PC: 10 # the number of principle components for your analysis
  nFeatLow: 200 # the lower level of cut-off for your number of features (aka genes in a cell)
  nFeatHigh: 10000 # the higher level cut-off of genes in a cell
  MThigh: 15 # the highest percent mitochondrial reads in a cell that you want to allow (to remove dying/dead cells)
  nCountHigh: 20000 # the high cutoff for total counts in a cell
  nVar: 2000 # the number of variable genes 
  GOI: "NCAM1" # the gene you are interested in seeing plotted
  GOI2: "EOMES" # the 2nd gene you are interested in plotted
  dir: "/Users/Jennifer.a.foltz/Library/CloudStorage/Box-Box/Foltz_Lab/Members/JAF/Presentations/BFX_workshop2025/Day1" # the file path to where your sample is located, and where you would like the output (html) file to be saved, and the seurat file in rds format
      
---
 Once you set the above parameters, click Knit in R studio.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library("Seurat")
library("ggplot2")
library("cowplot")
library("dplyr")
library("Matrix")
library("viridis")
```
Download dataset from here: https://www.10xgenomics.com/datasets/integrated-gex-totalseq-c-and-bcr-analysis-of-chromium-connect-generated-library-from-10k-human-pbmcs-2-standard; click the "Output and supplemental files" tab near the center of the page, scroll down, and choose "Gene Expression - Feature / cell matrix (per-sample)". Be sure to move these files into the directory you specified above.


```{r}
setwd(params$dir)

object.data <- Read10X("sample_feature_bc_matrix")

```
Here you are reading in the output from cellranger. This matrix is all droplets deemed to be cells by cellranger
If you have questions about functions and the arguments (e.g. the parameters) that can be passed to the function type ?nameoffunction and press enter and the right side of the Rstudio screen will display your options
```{r}
object <- CreateSeuratObject(object.data$`Gene Expression`, project = "object") 
```
Here you are creating a Seurat "object" which is the class of data that Seurat works with, feel free to call the project something besides object but it must be in quotes
```{r}
object[["percent.mt"]] = PercentageFeatureSet(object, pattern = "^MT-", assay = "RNA")
```
Here you assign this feature/gene set to the metadata, in this case, we want to know what percentage of counts are in mitochondrial genes as a way to look at dying cells.
```{r, fig.fullwidth=TRUE}
VlnPlot(object, features = c("percent.mt", "nFeature_RNA", "nCount_RNA"), pt.size = 0.1) 
```


Here you are plotting a QC plot.
Now choose the parameters that you think are the best for QC filtering, this will remove those cells from your Seurat object, and replot the qc thresholds
```{r, fig.width=7.5,fig.fullwidth = TRUE}
object <- subset(object, nFeature_RNA > params$nFeatLow & percent.mt < params$MThigh & nFeature_RNA < params$nFeatHigh & nCount_RNA < params$nCountHigh) # just examplenumbers, but replace with what you want

VlnPlot(object, features = c("percent.mt", "nFeature_RNA", "nCount_RNA"), pt.size = 0.1) 

```


Now, let's normalize the cell for different sequencing depths using lognormalization and find variable features, feel free to play with this setting to see how it changes your data
```{r}
object <- NormalizeData(object = object, normalization.method = "LogNormalize", scale.factor = 10000, assay = "RNA") 

object <- FindVariableFeatures(object = object, selection.method = 'vst', nfeatures = params$nVar, mean.cutoff = c(0.1, 8), dispersion.cutoff = c(1, Inf), assay = "RNA")
```

Here you are reading in lists of cell cycle genes in order to regress out the effect of cell cycle which can drive clustering more than cell lineages. With this function you are scoring each cell for cell cycle signatures which you should plot later to visualize whether the regression in scaledata removed these as a driver of clustering which we will discuss further in the future. 

ScaleData by default scales only the variable genes, if you want to scale all of them - e.g. to make a heatmap in Seurat you should set features = rownames(object) within the function
```{r}
s.genes <- cc.genes.updated.2019$s.genes 
g2m.genes <- cc.genes.updated.2019$g2m.genes
object <- CellCycleScoring(object=object, s.features=s.genes, g2m.features=g2m.genes)  
object <- ScaleData(object = object, vars.to.regress = c("S.Score","G2M.Score"), verbose = TRUE)
```

We will talk more about PCA and clustering in the future but for now run this code so you can get a feel for the data
```{r, fig.width = 7.5, fig.height= 5}
object <- RunPCA(object, npcs = 50, assay = "RNA")
object <- JackStraw(object, num.replicate = 100, dims = 30)

object <- ScoreJackStraw(object, dims = 1:30)
JackStrawPlot(object, dims = 1:30)

ElbowPlot(object)
```


It can be helpful to visualize the genes driving the PCs, and to be sure we are not excluding any populations of cells we are interested in that may account for only a small portion of the variability in the overall dataset (e.g. tumor clone, dendritic cells)

```{r, fig.width = 15, fig.height= 30, fig.fullwidth=TRUE}
DimHeatmap(object, dims = 1:12, balanced = TRUE, cells = 500)

DimHeatmap(object, dims = 13:24, balanced = TRUE, cells = 500)

DimHeatmap(object, dims = 25:36, balanced = TRUE, cells = 500)
```


Based on the above, how many PCs do you think are informative?
```{r, fig.fullwidth=TRUE}
object <- FindNeighbors(object, dims = 1:params$PC)

object <- FindClusters(object, resolution = params$res)

object <- RunTSNE(object, dims = 1:params$PC)

DimPlot(object, label = TRUE)

object <- RunUMAP(object, dims = 1:params$PC)

DimPlot(object, label = TRUE) # plot your UMAP with clusters visualized
```



Use feature plot here to visualize different markers, I have written some out as example but plot your favorite ones to see where they are located and whether expression is as you would expect
```{r, fig.width = 15, fig.height= 25, fig.fullwidth=TRUE}
FeaturePlot(object, features = c("nCount_RNA", "nFeature_RNA", "percent.mt", "GNLY", "TRG-AS1", "JCHAIN", "CD3G", "CD79A", "CD34", "LYZ", "HBA1", "GZMK", "S.Score", "G2M.Score", "PPBP", "MKI67", "CD3D", "TRAC", "CD8A", "CD4", "TRAV1-2"), ncol = 3, cols = c("lightgrey", "red"))
```
Now, let's dive into FeaturePlot and look at some ways data can be enhanced or altered.
The 1st plot shows the default FeaturePlot settings, in this example, the expression of a gene is shown as a scale and each cell's expression has no impact on where the cell is located. E.g. recall that in a UMAP/tsne/PCA cells can lie on top of each other in the dimensionality reduction.
In the 2nd plot, we tell Seurat, we want the cells that express our gene to be placed ON TOP of the other cells if they are buried.
In the 3rd plot, we give Seurat 5 colors to use (but this will be done with any colors > 2), and Seurat will no longer scale the data exactly but bins the data into values ranging from 1-5 (blue to red)
In the 4th plot, say we want a scaled expression not binned but we don't want the data necessarily binned, so we use viridis a color palette
In the 5th plot, we want to see how 2 genes are co-expressed, so we pass blend = TRUE, which will give us a new legend of single and double positive cells by color scale.
```{r, fig.fullwidth = TRUE, fig.width = 8}
FeaturePlot(object, features = params$GOI, cols = c("lightgrey", "red"))
FeaturePlot(object, features = params$GOI, cols = c("lightgrey", "red"), order = TRUE)
FeaturePlot(object, features = params$GOI, cols = c("blue", "green", "yellow2", "orange", "red"))
FeaturePlot(object, features = params$GOI) + scale_color_viridis()
FeaturePlot(object, features = c(params$GOI, params$GOI2), blend = TRUE)
```

Now, we want to see how PC 1 and PC 2 correspond to the umap layout. This can be helpful to look at larger patterns of data than clustering
```{r}
FeaturePlot(object, features = c("PC_1", "PC_2"))

```


Now, we want to find markers of each cluster, Seurat gives you multiple ways of doing this, 
You can use the FindAllMarkers function which is very convenient for finding markers of clusters; however, under the hood this function is a 1 cluster vs. all other clusters, so can miss subtle signals or miss genes that are conserved across multiple clusters but perhaps still important.
FindMarkers on the other hand, does pairwise comparisons which is more sensitive but less convenient
Note, that return.thresh which calls a p-value threshold calls it for unadjusted p-value, not adjusted which we definitely want to use,
FindMarkers on the other hand does not allow you to pass a p-value threshold so you will need to subset your markers afterwards
```{r}
DefaultAssay(object)
markers <- FindAllMarkers(object, min.pct = 0.25, logfc.threshold = 0.25, return.thresh = 0.05, only.pos= TRUE)
sigmarkers <- markers[markers$p_val_adj < 0.05,] # here we subset our markers for only the significant ones with adjusted p-value

```
Next, we use dplyr, to call some convenient functions to make better sense of our data, for example, we can ask dplyr to give us the top 5 markers for each cluster in our dataset and to determine this using the log2FC
Then, we can subset this list to look at what are those markers
We can also use this new dataframe to choose the markers to plot, a convenient function to limit the number of plots to inspect to determine whether our clustering is sufficient & we do want to do this, don't ever cluster blindly.
```{r, fig.fullwidth = TRUE}
top5 <- sigmarkers %>% group_by(cluster) %>% top_n(n=5, wt = avg_log2FC)
top5[top5$cluster == "0",]
FeaturePlot(object, features = unique(top5$gene)[1:6], ncol = 3, cols = c("lightgrey", "green4"))

```


Let's look at other convenient ways to visualize the data:

DoHeatmap within Seurat will give you your heatmap but in single cell form, where every column is a cell. Do note, that this pulls directly from the ScaleDate function so if you decide to remove some clusters in your heatmap, the data is scaled for the entire dataset, it is not rescaled for the heatmap creation. The advantage of DoHeatmap is it will give readers an idea of how many cells in a given group express a given gene, but it is more noisy that average expression heatmaps (see pheatmap or ComplexHeatmap)

DotPlot is a useful visualization that shows both scaled (based on the groups within the DotPlot itself; shown by color) and the amount of cells that express a given gene by the size of the dot

Violin Plots are a clear way to show genes, but take a fair amount of space, and can be difficult to visualize for genes that are lowly expressed.

RidgePlots are similar to a flow histogram, but are challenging for genes- better for protein data in my experience
```{r, fig.height = 8, fig.width=10, fig.fullwidth=TRUE}
DoHeatmap(object, features = unique(top5$gene)) + scale_fill_gradientn(colors = c("blue", "white", "red"))
DotPlot(object, features = unique(top5$gene)[1:20], scale.by = "size") + scale_colour_gradient2(low = "purple", mid = "grey", high = "pink")

VlnPlot(object, features = top5$gene[21:23], pt.size = 0, stack = TRUE, flip = TRUE)

RidgePlot(object, features = "CD3D")
```

```{r}
print(getwd())
setwd(params$dir)

saveRDS(object, file = "bfxworkshop.rds")

```


