# Week 18

## Single-cell RNAseq 

- [Lecture Recording](https://wustl.box.com/s/2q7e78fkm9bnz0lv4mnv3q696bd8k0r3)

- [Lecture Slides](BFXworkshop2025scRNADay1.pdf)

Presenter Dr. Jennifer Foltz


## Homework Assignments

To complete the homework assignment you will need to have R/Rstudio installed. You will not be able to open the Rmarkdown homeowrk file without this installation. In addition, we recommend installation of the following packages within R/Rstudio (note: RStudio may prompt you to install these when you open the Rmarkdown file) 
- Seurat
- ggplot2
- cowplot
- dplyr
- Matrix
- viridis

In the [Rmarkdown homework file](BFXworkshop_2025scRNA_Day1.Rmd), play around with the different parameters. Variables to modify are in the top portion and described within the Rmarkdown. Once you set the analysis parameters, click 'Knit' in RStudio to run the code. It will take about 5 minutes to run.  

Step through each of the inputs and outputs, and try modifying parameters:
 - what happens if you change the number of principal components?
 - try plotting a few other genes - which ones segregate in an "interesting" way?
 - what if you change the resolution of the clustering?

For-credit-students, Write a couple of sentences, explaining one of the parameters you changed, what it controls, and what the obvious effects on the downstream analyses were.  Email those to John, as usual.

As always, you can use the #bfx-workshop slack channel to ask questions.
